### PR TITLE
[RLlib] Replay Buffer API

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Review by gjoliver — COMMENTED** [review]
man I am a big fan of this change!! thank you so much!!
a quick question, did you make any logic updates while moving these code around.
thanks for doing this man!

File: rllib/utils/replay_buffers/mixin_replay_buffer.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        >>> buffer.add_batch(<A>)
+        >>> buffer.replay()
+        ... [<A>]
+        >>> buffer.add_batch(<B>)
+        >>> buffer.replay()
+        ... [<B>]
+    """
+
+    def __init__(self, capacity: int, storage_unit="timesteps",
+                 replay_ratio: float = 0.66):
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/mixin_replay_buffer.py:55]: I wanted to talk to someone about that. The alternative it to reorder the arguments and break the parent class' function signature. I'll be glad to change this however.

File: rllib/utils/replay_buffers/multi_agent_replay_buffer.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+            stat.update(
+                {"policy_{}".format(policy_id): replay_buffer.stats(
+                    debug=debug)}
+            )
+        return stat
+
+    @ExperimentalAPI
+    @DeveloperAPI
+    @override(ReplayBuffer)
+    def get_state(self) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/multi_agent_replay_buffer.py:297]: I have rewritten parts of the MixInReplayBuffer to support MultiAgent settings with prioritizations.
get_state and set_state are part of that. All other buffers had that method before.

**Comment by ArturNiederfahrenhorst** [comment]
> man I am a big fan of this change!! thank you so much!! a quick question, did you make any logic updates while moving these code around. thanks for doing this man!

Not yet, most of the code more of a sketch that Sven urged me to PR early on.
The plan is to write tests, fix, and add deprecation warnings to the old buffers before slowly moving logic from the old buffers to the new ones. Thanks for having a look at the code!

File: rllib/utils/replay_buffers/multi_agent_replay_buffer.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+        # Metrics.
+        self.add_batch_timer = TimerStat()
+        self.replay_timer = TimerStat()
+        self.update_priorities_timer = TimerStat()
+        self.num_added = 0
+
+        # Make externally accessible for testing.
+        global _local_replay_buffer
+        _local_replay_buffer = self
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/multi_agent_replay_buffer.py:125]: I'll have a look when writing the tests.

File: rllib/utils/replay_buffers/mixin_replay_buffer.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    MultiAgentBatch
+from ray.rllib.utils.annotations import DeveloperAPI, override, ExperimentalAPI
+from ray.rllib.utils.replay_buffers.multi_agent_replay_buffer import \
+    MultiAgentReplayBuffer
+from ray.rllib.utils.typing import PolicyID, SampleBatchType
+from ray.rllib.execution.buffers.replay_buffer import _ALL_POLICIES
+
+
+@ExperimentalAPI
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/utils/replay_buffers/mixin_replay_buffer.py:15]: Nit: Let's just use `@ExperimentalAPI` for now. Delete the `@DeveloperAPI` decorator.

File: rllib/agents/trainer.py [hunk 5]
```diff
@@ -359,6 +359,24 @@
         # "env_config": {...},
         # "explore": False
     },
+    # === Replay Buffer Settings ===
```
Comment (sven1977) [inline rllib/agents/trainer.py:362]: Awesome!

File: rllib/utils/replay_buffers/mixin_replay_buffer.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        replay_sequence_length: int = 1,
+        replay_burn_in: int = 0,
+        replay_zero_init_states: bool = True,
+        replay_ratio: float = 0.66,
+    ):
+        """Initializes MixInMultiAgentReplayBuffer instance.
+
+        Args:
+            capacity: Number of batches to store in total.
+            storage_unit (str):
```
Comment (sven1977) [inline rllib/utils/replay_buffers/mixin_replay_buffer.py:76]: Nit: fix comment.

File: rllib/utils/replay_buffers/mixin_replay_buffer.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        self.replay_proportion = None
+        if self.replay_ratio != 1.0:
+            self.replay_proportion = self.replay_ratio / (
+                1.0 - self.replay_ratio)
+
+        # Last added batch(es).
+        self.last_added_batches = collections.defaultdict(list)
+
+    @ExperimentalAPI
+    @DeveloperAPI
```
Comment (sven1977) [inline rllib/utils/replay_buffers/mixin_replay_buffer.py:142]: Same here, only `@ExperimentalAPI`.
There are a couple more instances of this below.

File: rllib/utils/replay_buffers/replay_buffer.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+        Returns:
+            The computer's networks name or an empty string, if the network
+            name could not be determined.
+        """
+        return platform.node()
+
+# TODO: What about this? Ask Sven?:
+# Visible for testing.
+# _local_replay_buffer = None
```
Comment (sven1977) [inline rllib/utils/replay_buffers/replay_buffer.py:206]: Let's get rid of it and see what happens. :)
This was solely for testing purposes anyways and is not used in any of our algorithms.

File: rllib/utils/replay_buffers/reservoir_buffer.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
+from typing import Any, Dict, Optional
```
Comment (sven1977) [inline rllib/utils/replay_buffers/reservoir_buffer.py:1]: Most awesome! :)

**Review by sven1977 — APPROVED** [review]
Thanks for the PR @ArturNiederfahrenhorst , this is really great. I feel like we should merge it already and add tests (and docs and all else) in a separate follow-up PR.

File: rllib/utils/replay_buffers/multi_agent_replay_buffer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/utils/replay_buffers/multi_agent_replay_buffer.py:123]: can you add a TODO here for now?
```
#TODO(artur) : Double check if this global variable is needed when adding unit tests.
```
thanks.

File: rllib/utils/replay_buffers/reservoir_buffer.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        # Update our timesteps counts.
+        self._num_timesteps_added += batch.count
+        self._num_timesteps_added_wrap += batch.count
+
+        if self._num_timesteps_added < self.capacity:
+            ReplayBuffer.add(self, batch)
+        else:
+            # Eviction of older samples has already started (buffer is "full")
+            self._eviction_started = True
+            idx = random.randint(0, self._num_add_calls - 1)
```
Comment (gjoliver) [inline rllib/utils/replay_buffers/reservoir_buffer.py:56]: mostly for my own curiosity, can you add a reference to the algorithm somehow?
I am a bit confused about randomly sampling from [0, num_add_calls].
over time, vast majority of the samples collected will get dropped with this approach?

File: rllib/utils/replay_buffers/reservoir_buffer.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+            if idx < self.capacity:
+                self._num_evicted += 1
+                self._evicted_hit_stats.push(self._hit_count[idx])
```
Comment (gjoliver) [inline rllib/utils/replay_buffers/reservoir_buffer.py:59]: you don't want to simply save the distribution _hit_count?
the list could grow pretty large over time, this is almost like a memory leak ... :)

**Review by gjoliver — COMMENTED** [review]
A couple of comments about Reservoir sampling. thanks again.

File: rllib/utils/replay_buffers/multi_agent_replay_buffer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/multi_agent_replay_buffer.py:123]: Sure, i'll make it part of my next PR!

File: rllib/utils/replay_buffers/reservoir_buffer.py (hunk 10 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/reservoir_buffer.py:56]: The algorithm itself is referenced in the class docstring.
Yes, a vast majority of samples will get dropped.
This is just me following the paper step by step.
Is far as I understand reservoir sampling, this holds true any implementation of reservoir buffers - if you want to sample all with equal probability from all seen samples any time, you would need infinite memory to hold all the samples. If you want to have a chance to sample your first experience after collecting a multiple of your buffer size in experiences, there needs to be a probability for new samples to get dropped.

I am curious and very open to discussion and a meeting about this.

File: rllib/utils/replay_buffers/reservoir_buffer.py (hunk 11 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/reservoir_buffer.py:59]: Hi! I have not looked into the implementation of WindowStat (=type(self._evicted_hit_stats)), and I will eventually, but WindowStat should drop old values as new ones are pushed, right? I was guessing only the last n values are of concern to such a statistic. The line occurs in the old ReplayBuffer class and I am yet to double check it.

**Pull request merged**